src/User.cpp

```diff
@@ -0,0 +1,120 @@
+    {
+        curDiff =  abs((*iter)->getLength() - _avgWatchableLen);
+        if (curDiff < minDiff && !isInHistory(**iter)) {
+            recommendedWatchable = *iter;
+            minDiff = curDiff;
+        }
+    }
+    return recommendedWatchable;
+}
+
+// Rerun Recommender User's methods implementation.
+RerunRecommenderUser::RerunRecommenderUser(const std::string& name):
+    User(name), _lastRecIdx(-1) {}
+
+Watchable* RerunRecommenderUser::getRecommendation(Session& s) {
+    if (_lastRecIdx == -1) {
+        return *(history.begin());
+    }
+    _lastRecIdx = (_lastRecIdx+1) % history.size();
+    return history[_lastRecIdx];
+}
+
+// Genre Recommender User's methods implementation.
+GenreRecommenderUser::GenreRecommenderUser(const std::string& name): User(name), _popularTagsMap(),
+                                                                        _popularTagsVector() {}
+
+void GenreRecommenderUser::updatePopularTags() {
+    std::vector<std::string> lastWatchedTags = history.back()->getTags();
+    //define iter, go through every tag, and update it in the data structure.
+    for (auto iter = lastWatchedTags.begin(); iter != lastWatchedTags.end(); iter++) {
+        // if tag doesn't exists in the map, add it
+        if (!_popularTagsMap.count(*iter)) {
+            _popularTagsVector.push_back(PopularTag(*iter));
+            _popularTagsMap[*iter] = &(_popularTagsVector.back());
+        }
+        else {
+            _popularTagsMap[*iter]->increaseCount();
+        }
+        // sorts by PopularTag's operator<()
+        std::sort(_popularTagsVector.begin(), _popularTagsVector.end());
+    }
+}
+
+
+void GenreRecommenderUser::addToHistory(Watchable& watchable) {
+    history.push_back(&watchable);
+    // update popular tags according to newly watched
+    updatePopularTags();
+}
+
+Watchable* GenreRecommenderUser::getRecommendation(Session &s) {
+    std::vector<Watchable*> content = s.getContent();
+    std::vector<Watchable*>::iterator contentIter;
+    std::vector<PopularTag>::iterator popularTagsIter;
+    // iterate over all user's popular tags, from most to least popular.
+    for (popularTagsIter = _popularTagsVector.begin(); popularTagsIter != _popularTagsVector.end();
+            popularTagsIter++) {
+        // iterate over the content, to check whether there is a Watchable containing this tag.
+        for (contentIter = content.begin(); contentIter != content.end(); contentIter++) {
+            // check that the watchable wasn't seen already
+            if (!isInHistory(**contentIter)) {
+                // extract the Watchable's tags vector.
+                const std::vector<std::string> &watchableTags = (*contentIter)->getTags();
+                // check if popular tag exist in Watchable's tags vector
+                if (std::find(watchableTags.begin(), watchableTags.end(), popularTagsIter->getName())
+                    != watchableTags.end()) {
+                    return *contentIter;
```

Comment:
Someone said in the whatsapp group that there should only be one return in each function.
Probably worth checking because it will be pretty easy to fix

Reply:
by course guidelines? I know it's like this for intro, but for SPL?! COME ON!!!